src/modules/voxelformat/GLTFFormat.cpp

```diff
@@ -0,0 +1,426 @@
+/**
+ * @file
+ */
+
+#include "GLTFFormat.h"
+#include "app/App.h"
+#include "core/Color.h"
+#include "core/Log.h"
+#include "core/Var.h"
+#include "core/concurrent/Lock.h"
+#include "core/concurrent/ThreadPool.h"
+#include "voxel/CubicSurfaceExtractor.h"
+#include "voxel/IsQuadNeeded.h"
+#include "voxel/MaterialColor.h"
+#include "voxel/Mesh.h"
+#include "voxel/VoxelVertex.h"
+#include <SDL_timer.h>
+#include <strstream>
+
+#define TINYGLTF_IMPLEMENTATION
+#include "external/tiny_gltf.h"
+
+namespace voxel {
+
+bool GLTFFormat::saveMeshes(const Meshes &meshes, const core::String &filename, io::SeekableWriteStream &stream,
+							const glm::vec3 &scale, bool quad, bool withColor, bool withTexCoords) {
+	return false;
+}
+
+void GLTFFormat::processGltfNode(tinygltf::Model &m, tinygltf::Node &node, tinygltf::Scene &scene,
+								 voxel::SceneGraphNode &graphNode, Stack &stack) {
+	node.name = graphNode.name().c_str();
+	const int idx = (int)m.nodes.size();
+
+	m.nodes.push_back(node);
+
+	if (stack.back().second != -1) {
+		m.nodes[stack.back().second].children.push_back(idx);
+	} else {
+		scene.nodes.push_back(idx);
+	}
+
+	stack.pop();
+
+	auto nodeChidren = graphNode.children();
+
+	for (int i = (int)nodeChidren.size() - 1; i >= 0; i--) {
+		stack.push_back(std::pair(nodeChidren[i], idx));
+	}
+}
+
+bool GLTFFormat::saveMeshes(const SceneGraph &sceneGraph, const Meshes &meshes, const core::String &filename,
+							io::SeekableWriteStream &stream, const glm::vec3 &scale, bool quad, bool withColor,
+							bool withTexCoords) {
+
+	const uint8_t UNSIGNED_SHORT_BYTES = 2;
+	const uint8_t FLOAT_BYTES = 4;
+
+	const voxel::Palette &palette = voxel::getPalette();
+	// 1 x 256 is the texture format that we are using for our palette
+	const float texcoord = 1.0f / (float)palette.colorCount;
+	// it is only 1 pixel high - sample the middle
+	unsigned char uvYVal[4] = {0, 0, 0, 63};
+
+	core::String palettename = core::string::stripExtension(filename);
+	palettename.append(".png");
+
+	const core::String &ext = core::string::extractExtension(filename);
+	const bool writeBinary = ext == "glb";
+	// const bool embedImages = false; // TODO research
+	// const bool embedBuffers = true; // TODO research
+	const bool prettyPrint = true;
+
+	tinygltf::TinyGLTF gltf;
+	tinygltf::Model m;
+
+	// Define the asset. The version is required
+	tinygltf::Asset asset;
+	asset.version = "2.0";
+	asset.generator = app::App::getInstance()->appname().c_str();
```

Comment:
we should include the version here, too

Reply:
see engine-config.h

Reply:
https://github.com/mgerhardy/vengi/commit/b1868ce681a572222ac62bb6450a36ed16ccf382